src/Control/Effect/Choose.hs

```diff
@@ -40,3 +44,26 @@ some a = (:) <$> a <*> many a
 -- | One or more, returning a 'NonEmpty' list of the results.
 some1 :: (Carrier sig m, Member Choose sig) => m a -> m (NonEmpty a)
 some1 a = (:|) <$> a <*> many a
+
+
+-- | Nondeterministically choose an element from a 'Foldable' collection.
+-- This can be used to emulate the style of nondeterminism associated with
+-- programming in the list monad:
+-- @
+--   pythagoreanTriples = do
+--     a <- oneOf [1..10]
+--     b <- oneOf [1..10]
+--     c <- oneOf [1..10]
+--     guard (a^2 + b^2 == c^2)
+--     pure (a, b, c)
+-- @
+oneOf :: (Foldable t, Carrier sig m, Member Choose sig, Member Empty sig) => t a -> m a
+oneOf = getChoosing . foldMap (Choosing . pure)
```

Comment:
This line of code took me on a journey. Specifically, the layout of this code reminded me of a quirk I found reading the `Data.Foldable` source code a long time ago: many of their core operations look like this:

```haskell
-- | Determines whether any element of the structure satisfies the predicate.
any :: Foldable t => (a -> Bool) -> t a -> Bool
any p = getAny #. foldMap (Any #. p)
```

Note the presence of that `#.` operator, as opposed to the more standard use of `.` `#.` is defined in Data.Functor.Utils, where it has a [long comment that I will not excerpt](http://hackage.haskell.org/package/base-4.12.0.0/docs/src/Data.Functor.Utils.html#%23.). The upshot of this is that there is a slight inefficiency associated with `g . f` constructions _if_ `g` is operationally equivalent to `id` at runtime (i.e. a newtype constructor or deconstructor). So we can just `coerce f` and we get something that doesn’t cause any further eta-expansion.

There exists a version of this `#.` operator for public consumption in the `Data.Profunctor.Unsafe` module¹ (note that despite the name of its enclosing module, this operator is _not_ implemented with `unsafeCoerce`—it’s all checked via `Coercible` constraints.)

I asked myself “well, I wonder if this shows up in the Core.” So I added the `dump-core` GHC plugin and found that the answer is _yes_! Here’s a picture of the Core generated from this PR as it stands:

<img width="731" alt="image" src="https://user-images.githubusercontent.com/2498/64755835-73218580-d4fa-11e9-9257-ff073644c568.png">

If we pull in `Data.Profunctor.Unsafe`, we can write this as so: 

```haskell
oneOf :: (Foldable t, Carrier sig m, Member Choose sig, Member Empty sig) => t a -> m a
oneOf = getChoosing #. foldMap (Choosing #. pure)
```

We then get substantially different Core (N.B.: this is all at `-O1`):

<img width="691" alt="image" src="https://user-images.githubusercontent.com/2498/64755879-a49a5100-d4fa-11e9-8058-405ce5a50101.png">

I can’t claim to be an expert at reading Core, but it sure looks like the second one is more efficient than the first. The `foldMap` seems to have been inlined away; `oneOf` becomes a single `let` and a call to `$woneOf`, which is a pretty straightforward call into the `$fMonoidChoosing` dictionary (?), which is unchanged between versions. In contrast, the first version has a `foldMap`, as well as some let-bindings that look pretty redundant to me.

I have not benchmarked this at all. But what do you think? Am I off on this? Am I reading this Core wrong? Should we pull in a `profunctors` dependency, or should we just implement our own `#.` with `coerce`? Put another way: _how weird do we wanna get?_

¹ Yes, Kmett and co. figured out that `#.` can be generalized to profunctors other than `->`. _Obviously._